pytools/mpi.py

```diff
@@ -29,3 +31,184 @@ def run_with_mpi_ranks(py_script, ranks, callable_, args=(), kwargs=None):
+        for name, value in param_dict.items():
+            if name == "num_tasks":
+                mpi_command += ["-n", str(value)]
+            else:
+                raise MPIExecutorParamError(name)
+        mpi_command += command
+        return mpi_command
+
+    def __call__(self, command, exec_params=None):
+        mpi_command = self.get_mpi_command(command, exec_params)
+        import subprocess
+        return subprocess.call(" ".join(mpi_command), shell=True)
+
+
+class SlurmMPIExecutor(MPIExecutor):
+    """Executor for Slurm-based platforms."""
+    def get_mpi_command(self, command, exec_params=None):
+        mpi_command = ["srun"]
+        param_dict = {}
+        if exec_params is not None:
+            param_dict = exec_params.param_dict
+        for name, value in param_dict.items():
+            if name == "num_tasks":
+                mpi_command += ["-n", str(value)]
+            elif name == "num_nodes":
+                mpi_command += ["-N", str(value)]
+            elif name == "tasks_per_node":
+                mpi_command += [f"--ntasks-per-node={value}"]
+            else:
+                raise MPIExecutorParamError(name)
+        mpi_command += command
+        return mpi_command
+
+    def __call__(self, command, exec_params=None):
+        mpi_command = self.get_mpi_command(command, exec_params)
+        import subprocess
+        return subprocess.call(" ".join(mpi_command), shell=True)
+
+
+class LCLSFMPIExecutor(MPIExecutor):
+    """Executor for Livermore wrapper around IBM LSF."""
+    def get_mpi_command(self, command, exec_params=None):
+        mpi_command = ["lrun"]
+        param_dict = {}
+        if exec_params is not None:
+            param_dict = exec_params.param_dict
+        for name, value in param_dict.items():
+            if name == "num_tasks":
+                mpi_command += ["-n", str(value)]
+            elif name == "num_nodes":
+                mpi_command += ["-N", str(value)]
+            elif name == "tasks_per_node":
+                mpi_command += ["-T", str(value)]
+            elif name == "gpus_per_task":
+                mpi_command += ["-g", str(value)]
+            else:
+                raise MPIExecutorParamError(name)
+        mpi_command += command
+        return mpi_command
+
+    def __call__(self, command, exec_params=None):
+        mpi_command = self.get_mpi_command(command, exec_params)
+        import subprocess
+        return subprocess.call(" ".join(mpi_command), shell=True)
+
+
+def make_mpi_executor(executor_type_name):
+    """
+    Returns an instance of a class derived from :class:`MPIExecutor` given an
+    executor type name as input.
+
+    :arg executor_type_name: The executor type name. Can be one of `'basic'`,
+        `'slurm'`, or `'lclsf'`.
+    """
+    type_name_map = {
+        "basic": BasicMPIExecutor,
+        "slurm": SlurmMPIExecutor,
+        "lclsf": LCLSFMPIExecutor
+    }
+    return type_name_map[executor_type_name]()
```

Comment:
Would it make sense to create some rudimentary auto-detection facility (which can be overridden by a documented env var)?